### PR TITLE
Add Bert modules

### PR DESCRIPTION
**Pull request by gpengzhi**
1. Add `tensorflow` to `requirements` since we need to load `tf` checkpoint file for BERT modules.
2. Add `BertBase`, `bert_utils`, and `bert_utils_test`.
3. Add `BertEncoder` and the corresponding test.
4. Add `BertClassifierV2` and the corresponding tests for `BertClassifierV1` and `BertClassifierV2`.
5. Add `bert_classifier_main_v2` in `bert` exmaple.

### CODE REVIEW

File: docs/code/modules.rst [hunk 1]
```diff
@@ -204,3 +214,17 @@ Networks
...
     :members: default_hparams,forward,append_layer,has_layer,layer_by_name,layers_by_name,layers,layer_names
+
+Bert
+=========
+
+.. spelling::
+    pretrained
+    pooler
+    hparams
+    bert
```
Comment (huzecong) [inline docs/code/modules.rst]: - The correct spelling for "*pretrained*" is actually "*pre-trained*".
- "*hparams*" should not be a word -- either use "*hyperparameters*" or `` `hparams` ``.
- Similarly, use "*BERT*" instead of "*bert*". The former would probably not trigger a warning because it is considered an acronym.

File: examples/bert/bert_classifier_main_V2.py [hunk 2]
```diff
@@ -0,0 +1,257 @@
...
+    # Loads data
+    num_train_data = config_data.num_train_data
+
+    # Builds BERT
+    hparams = {
+        'clas_strategy': 'cls_time'
+    }
+    model = BertClassifierV2(hparams=hparams)
+    if torch.cuda.is_available():
+        model = model.cuda()
```
Comment (huzecong) [inline examples/bert/bert_classifier_main_V2.py]: The following might be a better style:
https://github.com/asyml/texar-pytorch/blob/fc99e0db92666caa947aba96e4c950db3aeb70a2/examples/transformer/transformer_main.py#L61-L65

This way you don't have to check for `cuda.is_available()` everywhere, you can just call `.to(device)`. Also, you can pass `device` to `RecordData` constructor (or call `.to` after construction) to automatically move input to the appropriate device.

File: examples/bert/bert_classifier_main_V2.py [hunk 3]
```diff
@@ -0,0 +1,257 @@
...
+
+    @torch.no_grad()
+    def _eval_epoch():
+        """Evaluates on the dev set.
+        """
+        iterator.switch_to_dataset("eval")
+        model.eval()
+        cum_acc = 0.0
+        cum_loss = 0.0
+        nsamples = 0
```
Comment (huzecong) [inline examples/bert/bert_classifier_main_V2.py]: We can use `texar.utils.AverageRecorder` instead of manually keeping track of these.

File: texar/modules/__init__.py [hunk 4]
```diff
@@ -15,6 +15,7 @@
 Modules of Texar library module.
 """
 
+from texar.modules.berts import *
```
Comment (huzecong) [inline texar/modules/__init__.py]: Can we use some other name to replace `berts`? This sounds really strange...

I suggest calling this package along the lines of `representations` or `pretrained`. If we plan to add ELMo, XLNet and friends, they could go here as well.

File: texar/modules/berts/bert_base.py [hunk 5]
```diff
@@ -0,0 +1,91 @@
...
+            pre-trained models will be cached. If `None` (default),
+            a default directory will be used.
+        hparams (dict or HParams, optional): Hyperparameters. Missing
+            hyperparameter will be set to default values. See
+            :meth:`default_hparams` for the hyperparameter structure
+            and default values.
+    """
+
+    def __init__(self,
+                 pretrained_model_name=None,
```
Comment (huzecong) [inline texar/modules/berts/bert_base.py]: Please add type annotations here.

File: texar/modules/berts/bert_base.py [hunk 6]
```diff
@@ -0,0 +1,91 @@
...
+                 cache_dir=None,
+                 hparams=None):
+        ModuleBase.__init__(self, hparams)
```
Comment (huzecong) [inline texar/modules/berts/bert_base.py]: Prefer using `super().__init__(hparams)` instead of explicitly writing the base class name. This could lead to problems in case we change the class hierarchies.

File: texar/modules/berts/bert_utils.py [hunk 7]
```diff
@@ -0,0 +1,299 @@
...
+    'bert-base-multilingual-cased':
+        _BERT_PATH + "2018_11_03/multilingual_L-12_H-768_A-12.zip",
+    'bert-base-chinese':
+        _BERT_PATH + "2018_11_03/chinese_L-12_H-768_A-12.zip",
+}
+
+
+def init_bert_checkpoint(model: nn.Module, cache_dir: str):
+    r"""
+    Initializes BERT model parameters from a checkpoint provided by Google.
```
Comment (huzecong) [inline texar/modules/berts/bert_utils.py]: Although either is okay, we don't have newlines after `r"""` for all other docstrings. It might be better to be consistent.

File: texar/modules/classifiers/bert_classifiers.py [hunk 8]
```diff
@@ -312,3 +334,237 @@ def forward(self, inputs: torch.Tensor,  # type: ignore
...
+            "dropout": 0.1,
+            "name": "bert_classifier"
+        })
+        return hparams
+
+    def forward(self,  # type: ignore
+                inputs: torch.Tensor,
+                sequence_length: Optional[torch.LongTensor] = None,
+                segment_ids: Optional[torch.LongTensor] = None) -> \
+            Tuple[torch.Tensor, torch.LongTensor]:
```
Comment (huzecong) [inline texar/modules/classifiers/bert_classifiers.py]: Similarly, we wrap before the arrow `->`.

File: texar/modules/encoders/bert_encoders.py [hunk 9]
```diff
@@ -0,0 +1,342 @@
...
+            pre-trained models will be cached. If `None` (default),
+            a default directory will be used.
+        hparams (dict or HParams, optional): Hyperparameters. Missing
+            hyperparameter will be set to default values. See
+            :meth:`default_hparams` for the hyperparameter structure
+            and default values.
+    """
+
+    def __init__(self,
+                 pretrained_model_name=None,
```
Comment (huzecong) [inline texar/modules/encoders/bert_encoders.py]: Please add type annotations.

File: texar/modules/encoders/bert_encoders.py [hunk 10]
```diff
@@ -0,0 +1,342 @@
...
+
+        batch_size = inputs.shape[0]
+        pos_length = torch.ones(batch_size, dtype=torch.int64) * inputs.shape[1]
+        pos_embeds = self.position_embedder(sequence_length=pos_length)
+
+        inputs_embeds = word_embeds + segment_embeds + pos_embeds
+
+        if sequence_length is None:
+            sequence_length = torch.ones(batch_size, dtype=torch.int64) \
+                              * inputs.shape[1]
```
Comment (huzecong) [inline texar/modules/encoders/bert_encoders.py]: Prefer `torch.full` to this.

File: texar/modules/berts/bert_utils.py [hunk 11]
```diff
@@ -0,0 +1,299 @@
...
+    'bert-base-multilingual-uncased':
+        _BERT_PATH + "2018_11_23/multi_cased_L-12_H-768_A-12.zip",
+    'bert-base-multilingual-cased':
+        _BERT_PATH + "2018_11_03/multilingual_L-12_H-768_A-12.zip",
+    'bert-base-chinese':
+        _BERT_PATH + "2018_11_03/chinese_L-12_H-768_A-12.zip",
+}
+
+
+def init_bert_checkpoint(model: nn.Module, cache_dir: str):
```
Comment (ZhitingHu) [inline texar/modules/berts/bert_utils.py]: Can users choose to disable `init_bert_checkpoint` call in the module and instead load the checkpoint by themselves?

Comment (gpengzhi) [reply texar/modules/berts/bert_utils.py]: When the users create `model = BertClassifier(...)`, they could set `pretrained_model_name=None` and `pretrained_model_name` in `hparams` to be `None` so that `BertClassifier` won't load any pre-trained model. The users then could use `torch.load(checkpoint_file)` and `model.load_state_dict(...)` to load their own checkpoint file.

File: texar/modules/__init__.py (hunk 4 above)
Comment (ZhitingHu) [reply texar/modules/__init__.py]: yes, `pretrained` looks like a good name. @gpengzhi can you update Texar-TF as well

File: examples/bert/bert_classifier_main_V2.py [hunk 12]
```diff
@@ -0,0 +1,257 @@
...
+                                              reduction='mean')
+            else:
+                loss = F.cross_entropy(logits.view(-1, model.num_classes),
+                                       labels.view(-1),
+                                       reduction='mean')
+
+            loss.backward()
+            optim.step()
+            scheduler.step()
+            step = scheduler.last_epoch
```
[the remark this answers is not in the document]
Comment (gpengzhi) [reply examples/bert/bert_classifier_main_V2.py]: Maybe we can discuss this part after the pipeline meeting tomorrow.

**Comment by gpengzhi** [comment]
1. polish `bert` example. The example is implemented by the lasted `bert` modules now.
2. move `BertAdam` to `texar.core`
3. delete `BertClassifierV1`.

File: examples/bert/bert_classifier_main_V2.py [hunk 13]
```diff
@@ -0,0 +1,257 @@
...
+            dis_steps = config_data.display_steps
+            if dis_steps > 0 and step % dis_steps == 0:
+                logging.info("step: %d; loss: %f", step, loss)
+
+            eval_steps = config_data.eval_steps
+            if eval_steps > 0 and step % eval_steps == 0:
+                _eval_epoch()
+
+    @torch.no_grad()
+    def _eval_epoch():
```
[the remark this answers is not in the document]
Comment (gpengzhi) [reply examples/bert/bert_classifier_main_V2.py]: I feel it here is a little hard to reduce the code redundancy. Maybe we can discuss this part as well tomorrow.

File: texar/modules/classifiers/bert_classifiers.py [hunk 14]
```diff
@@ -80,235 +91,172 @@ def __init__(self, hparams=None):
...
+            logits = self._dropout_layer(logits)
+            logits = self._logits_layer(logits)
 
-        num_classes = self._num_classes
-        is_binary = num_classes == 1
-        is_binary = is_binary or (num_classes <= 0 and logits.shape[-1] == 1)
+        # Compute predictions
+        self.is_binary = self.num_classes == 1
+        self.is_binary = self.is_binary or (self.num_classes <= 0 and
+                                            logits.shape[-1] == 1)
```
Comment (huzecong) [inline texar/modules/classifiers/bert_classifiers.py]: Why can't we check whether classification is binary or not in `__init__`?

File: texar/modules/pretrained/__init__.py [hunk 15]
```diff
@@ -0,0 +1,21 @@
...
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+"""
+Modules of Texar library bert.
+"""
+
+# pylint: disable=wildcard-import
```
Comment (huzecong) [inline texar/modules/pretrained/__init__.py]: This Pylint directive is probably unnecessary.

File: texar/modules/classifiers/bert_classifiers.py [hunk 16]
```diff
@@ -17,60 +17,71 @@
...
+            `bert-large-cased`, `bert-base-multilingual-uncased`,
+            `bert-base-multilingual-cased`, `bert-base-chinese`.
+            If `None`, will use the model name in :attr:`hparams`.
+        cache_dir (optional): the path to a folder in which the
+            pre-trained models will be cached. If `None` (default),
+            a default directory will be used.
         hparams (dict or HParams, optional): Hyperparameters. Missing
-            hyperparameters will be set to default values. See
-            :meth:`default_hparams` for the hyper-parameter structure
+            hyperparameter will be set to default values. See
```
Comment (huzecong) [inline texar/modules/classifiers/bert_classifiers.py]: This should actually be "hyperparameters".

File: texar/modules/classifiers/bert_classifiers.py [hunk 17]
```diff
@@ -80,235 +91,172 @@ def __init__(self, hparams=None):
...
-            {
-                # (1) Different modules for sentence encoder
-                # (2) Additional hyperparameters
-                "num_classes": 2,
-                "logit_layer_kwargs": None,
-                "clas_strategy": "cls_time",
-                "max_seq_length": None,
-                "dropout": 0.1,
-                "name": "bert_classifier"
+        {
```
Comment (huzecong) [inline texar/modules/classifiers/bert_classifiers.py]: Is the indentation wrong here?

File: texar/modules/classifiers/bert_classifiers.py [hunk 18]
```diff
@@ -80,235 +91,172 @@ def __init__(self, hparams=None):
...
+            # (1) Same hyperparameters as in BertEncoder
+            ...
+            # (2) Additional hyperparameters
+            `num_classes`: 2,
```
Comment (huzecong) [inline texar/modules/classifiers/bert_classifiers.py]: If this is a code example, we should use quotes (`"`) instead of backticks (`` ` ``).

File: texar/modules/classifiers/bert_classifiers.py [hunk 19]
```diff
@@ -80,235 +91,172 @@ def __init__(self, hparams=None):
...
-               Maximum possible length of input sequences. Required if
-               `"clas_strategy"` is "all_time".
-
-           `"dropout"`: float
-               The dropout rate of the BERT encoder output.
-
-           `"name"`: str
-               Name of the classifier.
+        1. Same hyperparameters as in
+        :class:`~texar.modules.BertEncoder`.
```
Comment (huzecong) [inline texar/modules/classifiers/bert_classifiers.py]: Indentation here is incorrect as well.

File: texar/modules/classifiers/bert_classifiers.py [hunk 20]
```diff
@@ -80,235 +91,172 @@ def __init__(self, hparams=None):
...
+        See the :meth:`~texar.modules.BertEncoder.default_hparams`.
+        An instance of BertEncoder is created for feature extraction.
+
+        2. Additional hyperparameters:
+
+            `num_classes` : int
```
Comment (huzecong) [inline texar/modules/classifiers/bert_classifiers.py]: There should not be a space before the colon (`:`).

File: examples/bert/bert_classifier_main_V2.py (hunk 13 above)
[the remark this answers is not in the document]
Comment (ZhitingHu) [reply examples/bert/bert_classifier_main_V2.py]: Can we have a `_run_epoch` (like [here](https://github.com/asyml/texar/blob/93c6c1980a66683884798f33e67bc855137af28c/examples/sentence_classifier/clas_main.py#L61)) which does the common operations, and have `_train_opech` and `_eval_epoch` wrapping over `_run_epoch`? We can chat offline if needed

[the remark this answers is not in the document]
Comment (gpengzhi) [reply examples/bert/bert_classifier_main_V2.py]: I feel that merging `_train_epoch` and `_eval_epoch` will reduce the code readability. For example, `_train_epoch` has some optimization-related code (`optim.zero_grad()`, `loss.backward()`, ...) in the loop, and we have `@torch.no_grad()` and some other eval-related code in `_eval_epoch`. The code here might have some code redundancy, but  I feel it is easy to read for the audience. Please let me know your thought.

[the remark this answers is not in the document]
Comment (ZhitingHu) [reply examples/bert/bert_classifier_main_V2.py]: I think the following part can be shared (it's extact the same in `_train_epoch` and `_eval_epoch`, right?):
https://github.com/asyml/texar-pytorch/blob/97b597e1eb3706847458c878713e46f8c099aa8e/examples/bert/bert_classifier_main_V2.py#L134-L154

It's for evaluating loss. It's a common case that train and eval share this part, like:
```
Train:
    loss = evaluate_loss(train_batch)
    optimize(loss)

Eval:
   loss = evaluate_loss(dev_batch)
   average_loss(loss)
```

File: texar/modules/encoders/bert_encoders.py [hunk 21]
```diff
@@ -120,105 +120,105 @@ def default_hparams():
         .. code-block:: python
 
             {
-            'pretrained_model_name': 'bert-base-uncased',
-            'embed': {
-                'dim': 768,
-                'name': 'word_embeddings'
+            "pretrained_model_name": "bert-base-uncased",
```
Comment (huzecong) [inline texar/modules/encoders/bert_encoders.py]: This part is Python code, so it should be formatted as Python code. The contents of this bracket should be further indents. This also applied to the `BertClassifier` example.

File: texar/modules/classifiers/bert_classifiers.py [hunk 22]
```diff
@@ -100,36 +100,39 @@ def __init__(self,
...
+            "name": "bert_classifier"
             }
 
         Here:
 
         1. Same hyperparameters as in
-        :class:`~texar.modules.BertEncoder`.
-        See the :meth:`~texar.modules.BertEncoder.default_hparams`.
-        An instance of BertEncoder is created for feature extraction.
+            :class:`~texar.modules.BertEncoder`.
```
Comment (huzecong) [inline texar/modules/classifiers/bert_classifiers.py]: Following lines in a list should indent to the beginning of the first line, in this case, right after `1. `. So these lines should be indented with 3 spaces, not 4.

Comment (gpengzhi) [reply texar/modules/classifiers/bert_classifiers.py]: Sorry for all the docs-related error. I should have checked the generated `html` before submitting the PR.
